### PR TITLE
More type hints for `synapse.logging`

### PR DESCRIPTION
**Pull request by clokep**
`scopecontextmanager` is now fully typed and `opentracing` has some more types.

I have a follow-on to this that I'm still working through, typing some of the decorators causes many errors. 😢

### CODE REVIEW

File: synapse/logging/opentracing.py [hunk 1]
```diff
@@ -617,23 +617,25 @@ def set_operation_name(operation_name: str) -> None:
 
 
 @only_if_tracing
-def force_tracing(span=Sentinel) -> None:
+def force_tracing(span: "opentracing.Span" = Sentinel) -> None:  # type: ignore[assignment]
```
Comment (squahtx) [inline synapse/logging/opentracing.py:620]: Elsewhere in synapse we tend to define a dummy enum like so (search for `_Sentinel`):
```py
class _Sentinel(enum.Enum):
    # defining a sentinel in this way allows mypy to correctly handle the
    # type of a dictionary lookup.
    sentinel = object()
```
Then we can avoid the mypy complaints:
```suggestion
def force_tracing(span: Union["opentracing.Span", _Sentinel] = Sentinel) -> None:
```

File: synapse/logging/opentracing.py [hunk 2]
```diff
@@ -617,23 +617,25 @@ def set_operation_name(operation_name: str) -> None:
...
     Args:
         span: span to force tracing for. By default, the active span.
     """
     if span is Sentinel:
-        span = opentracing.tracer.active_span
-    if span is None:
+        span_to_trace = opentracing.tracer.active_span
+    else:
+        span_to_trace = span
+    if span_to_trace is None:
```
Comment (squahtx) [inline synapse/logging/opentracing.py:630]: Seems to imply that `span` can be `None`, which runs counter to the type hint? Do we ever pass a `None` in somewhere?

Comment (clokep) [reply synapse/logging/opentracing.py:630]: I think passing `None` in is invalid, but `opentracing.tracer.active_span` can be `None`? Will double check.

Comment (squahtx) [reply synapse/logging/opentracing.py:630]: Makes sense, I misread the code.

**Review by squahtx — APPROVED** [review]
lgtm!

**Pull request merged**